### ISSUE TITLE
Wrong hostname in pod work container for Windows nodes with containerd runtime

### ISSUE THREAD

**Issue by adelina-t** [kubernetes/kubernetes#82711]
Windows + containerd

When creating a pod, the expectation is that all containers have the hostname properly set to the name of the pod. 

However when using Windows nodes + containerd as a runtime, only the pause container in the pod has the hostname properly set, not the work container.

**How to reproduce it (as minimally and precisely as possible)**:

```
kubectl run test --image test --image e2eteam/busybox:1.29 -- sleep 1000

kubectl exec test -- ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : 3e3bc2daf987
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : default.svc.cluster.local

Ethernet adapter vEthernet (05dded9b071f5aa8b404783c6add3f9de58f2d9a6e88dcdf62c4c19f6039ae2d_cbr0):

   Connection-specific DNS Suffix  . : default.svc.cluster.local
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #3
   Physical Address. . . . . . . . . : 00-15-5D-44-4C-12
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::35f8:39fd:bb24:e58a%22(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.244.1.5(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.244.1.2
   DNS Servers . . . . . . . . . . . : 10.96.0.10
   NetBIOS over Tcpip. . . . . . . . : Disabled
   Connection-specific DNS Suffix Search List :
                                       default.svc.cluster.local
```

From the Windows node:

Run ctr.exe exec on the pause container for the pod, you can see the proper hostname set.

```
.\ctr.exe --namespace k8s.io task exec --exec-id 1010 05dded9b071f5aa8b404783c6add3f9de58f2d9a6e88dcdf62c4c19f6039ae2d ipconfig /all

Windows IP Configuration

   Host Name . . . . . . . . . . . . : test2-d7c794d76-mwkkw
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : default.svc.cluster.local

Ethernet adapter vEthernet (05dded9b071f5aa8b404783c6add3f9de58f2d9a6e88dcdf62c4c19f6039ae2d_cbr0):

   Connection-specific DNS Suffix  . : default.svc.cluster.local
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #3
   Physical Address. . . . . . . . . : 00-15-5D-44-4C-12
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::35f8:39fd:bb24:e58a%22(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.244.1.5(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.244.1.2
   DNS Servers . . . . . . . . . . . : 10.96.0.10
   NetBIOS over Tcpip. . . . . . . . : Disabled
   Connection-specific DNS Suffix Search List :
                                       default.svc.cluster.local
**Environment**:
- Kubernetes version (use `kubectl version`):
- Cloud provider or hardware configuration:
- OS (e.g: `cat /etc/os-release`):
- Kernel (e.g. `uname -a`):
- Install tools:
- Network plugin and version (if this is a network-related bug):
- Others:
```

Run the same on the work container for the pod, hostname is not properly set

```
PS C:\k> .\ctr.exe --namespace k8s.io task exec --exec-id 1010 79e55512f9a27fa888e176e7e9376dc6f9b52b5f3b4e307167420276fb92019a ipconfig /all 

Windows IP Configuration

   Host Name . . . . . . . . . . . . : 3e3bc2daf987
   Primary Dns Suffix  . . . . . . . :
   Node Type . . . . . . . . . . . . : Hybrid
   IP Routing Enabled. . . . . . . . : No
   WINS Proxy Enabled. . . . . . . . : No
   DNS Suffix Search List. . . . . . : default.svc.cluster.local

Ethernet adapter vEthernet (05dded9b071f5aa8b404783c6add3f9de58f2d9a6e88dcdf62c4c19f6039ae2d_cbr0):

   Connection-specific DNS Suffix  . : default.svc.cluster.local
   Description . . . . . . . . . . . : Hyper-V Virtual Ethernet Adapter #3
   Physical Address. . . . . . . . . : 00-15-5D-44-4C-12
   DHCP Enabled. . . . . . . . . . . : No
   Autoconfiguration Enabled . . . . : Yes
   Link-local IPv6 Address . . . . . : fe80::35f8:39fd:bb24:e58a%22(Preferred)
   IPv4 Address. . . . . . . . . . . : 10.244.1.5(Preferred)
   Subnet Mask . . . . . . . . . . . : 255.255.255.0
   Default Gateway . . . . . . . . . : 10.244.1.2
   DNS Servers . . . . . . . . . . . : 10.96.0.10
   NetBIOS over Tcpip. . . . . . . . : Disabled
   Connection-specific DNS Suffix Search List :
                                       default.svc.cluster.local
```

**Environment info**

containerd + l2bridge cni plugin. 
Windows 1809 10.0.17763.557

**Comment by ksubrmnn** [comment]
@nagiesek Do you know how this would happen? Is this a CNI thing? I also don't have context here as to why this would work differently than it would for Docker.

**Comment by nagiesek** [comment]
I don't think setting hostname is commonly supported through CNI. My expectation is that the hostname would be set as part of the CRI, but I'd have to dig around a bit to confirm.

**Comment by PatrickLang** [comment]
this causes failures:

```
[sig-network] [sig-windows] Networking Granular Checks: Pods should function for intra-pod communication: http
[sig-network] [sig-windows] Networking Granular Checks: Pods should function for intra-pod communication: udp
[sig-network] [sig-windows] Networking Granular Checks: Pods should function for node-pod communication: http
[sig-network] [sig-windows] Networking Granular Checks: Pods should function for node-pod communication: udp
```

**Comment by adelina-t** [comment]
> I don't think setting hostname is commonly supported through CNI. My expectation is that the hostname would be set as part of the CRI, but I'd have to dig around a bit to confirm.

That's what confuses me as well. Looking trough the code in dockershim, the hostname is set explicitly only on the sandbox container. I've checked, and the work container gets it's proper hostname only after it is started. I've put a long sleep in between create and start operations and indeed it is the case. So I assume there is an extra step happening at start where the hostname is updated, but I haven't figured out if CNI is involved ( I don't see how ).

**Comment by adelina-t** [comment]
Also causes these tests to fail:

```
[sig-apps] ReplicaSet should serve a basic image on each replica with a public image [Conformance] 
[sig-apps] ReplicationController should serve a basic image on each replica with a public image [Conformance]
```

**Comment by nagiesek** [comment]
@jterry75 apologies for dragging you into kubernetes, but do you have any suggestions/insights about this hostname issue?

**Comment by jterry75** [comment]
The hostname is not part of the container create config as it is inherited from the sandbox. But it looks like potentially the fix in Windows Containers is that we need to assign the hostname of the sandbox per container creation. This is different than what we do for linux where the shared namespace already has the single hostname. I think the fix will be from CRI

**Comment by adelina-t** [comment]
@jterry75  something as simple as: https://github.com/adelina-t/cri/commit/5d8e17860068d5d9e8276dab73bbcbbe4136a263 should work, however I still fell this is a temp fix. Dockershim doesn't explicitly set the hostname via the container config and yet it works after the container starts, which leads me to believe that there is an extra call ( CNI? ) that takes care of this.

**Comment by PatrickLang** [comment]
per @dineshgovindasamy his team is looking into it. One of the keys that needs to be set at container startup isn't set.

**Comment by jterry75** [comment]
@adelina-t - 100% correct that fix will work. But as @PatrickLang said we are trying to see if there is another bug at fault here before we create this workaround. Thanks!

**Comment by jterry75** [comment]
@adelina-t - On second thought I see no issues with just fixing this now as well. Here is the PR: https://github.com/containerd/cri/pull/1294

**Comment by PatrickLang** [comment]
@adelina-t  - this seems to be working the last two days with that change. Is this ready to close?

**Comment by PatrickLang** [comment]
@Random-Liu @adelina-t - I was mistaken and this isn't tracked as part of the network namespace. The fixes.

open question - is there any case where we wouldn't set this when creating the sandbox?
